lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,342 @@
+//===- GrandCentral.cpp - Ingest black box sources --------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// Implement SiFive's Grand Central transform.
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Pass Implementation
+//===----------------------------------------------------------------------===//
+
+namespace {
+namespace Component {
+enum class Type { Ground, Vector, Bundle };
+
+struct Info {
+  Type tpe;
+  StringRef description;
+  bool found = false;
+  int32_t width = -1;
+  uint32_t depth = 0;
```

Comment:
It looks like you're reencoding type information here.  Why not just capture the FIRRTL type instead of unpacking into a width/depth?

Reply:
Yes, it's re-encoding the type information. I had avoided that because neither the width nor the depth is known and has to be mutably updated and it seemed bad to keep creating `IntegerType` or `hw::UnpackedArrayType` that are thrown away because I can't mutably update their storage.

Is it bad to create types that are just thrown away like this?

It's a multi-step process to build this up. First, a circuit-level annotation defines the structure of the interface which says, "I have an element that is of type `Ground | Vector | Bundle`" (where a bundle is a nested interface). However, it doesn't say anything about the number of bits in the type or the width of a vector type. Second, the circuit is walked and I find the component that corresponds to a ground type or the components that make up a vector type.

I thought about this using the actual type, but discounted it for the reason above. 

I also tried to implement this by building out a new bunch of structs that could work with LLVM's RTTI, but didn't do the storage correctly and couldn't get it to work. (I then bailed on that because it seemed like overkill.)

What do you think?

Reply:
Ok, I didn't realize you had to do that processing.  I'm ok with keeping it this way if htere is a good reason like that.

There isn't a significant cost to throwing away integer and vector types, they're uniqued.